openhtf/util/data.py

```diff
@@ -138,3 +139,30 @@ def ConvertToBaseTypes(obj, ignore_keys=tuple()):
     obj = str(obj)
 
   return obj
+
+
+def TotalSize(obj):
+  """Returns the approximate total memory footprint an object."""
+  seen = set()
+  def sizeof(current_obj):
+    """Do a depth-first acyclic traversal of all reachable objects."""
+    if id(current_obj) in seen:
+      # A rough approximation of the size cost of an additional reference.
+      return struct.calcsize('P')
+    seen.add(id(current_obj))
+    size = sys.getsizeof(current_obj)
+
+    if isinstance(current_obj, dict):
+      size += sum(map(sizeof, itertools.chain.from_iterable(
+          current_obj.iteritems())))
+    elif (isinstance(current_obj, collections.Iterable) and
+          not isinstance(current_obj, basestring)):
+      size += sum(map(sizeof, iter(current_obj)))
+    elif isinstance(current_obj, records.RecordClass):
+      size += sizeof(current_obj.__slots__)
+      size += sum(map(sizeof, current_obj.optional_attributes.itervalues()))
```

Comment:
`optional_attributes` isn't either, only the data in the slots themselves (or else what's the point of `__slots__`). These three lines can simplify to:

```py
size += sum(sizeof(getattr(current_obj, attr)) for attr in current_obj.__slots__)
```